### ISSUE TITLE
Unsupported usage unit of Azure - 1 GB Hour/1 GB Second

### ISSUE THREAD

**Issue by HankLeo** [cloud-carbon-footprint/cloud-carbon-footprint#1189]
## Feature Suggestion

Hi,

I get below errors in my project when using cloud-carbon-footprint to calculate Azure carbon emissions.
```
2023-07-19T05:43:27.021Z [ConsumptionManagement] warn: Unsupported usage unit: 1 GB Second 
2023-07-19T05:43:27.043Z [ConsumptionManagement] warn: Unsupported usage unit: 1 GB Hour 
2023-07-19T05:43:27.044Z [ConsumptionManagement] warn: Unsupported usage unit: 1 GB Second 
2023-07-19T05:43:27.044Z [ConsumptionManagement] warn: Unsupported usage unit: 1 GB Second 
```

Some of our Azure billing usage like below:

resourceLocation | meterCategory | meterName | unitOfMeasure | consumedQuantity
-- | -- | -- | -- | --
EastUS | Functions | Standard Execution Time | 1 GB Second | 12.0185
CentralUS | Functions | Standard Execution Time | 1 GB Second | 180.455375
CentralUS | Functions | Standard Execution Time | 1 GB Second | 368.099625
CentralUS | Container Instances | Standard Memory Duration | 1 GB Hour | 2546.8697512514464
CentralUS | Container Instances | Standard Memory Duration | 1 GB Hour | 1488.0003837521679
CentralUS | Container Instances | Standard Memory Duration | 1 GB Hour | 0.165881712458333

But I found that in your [Azure ConsumptionTypes](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/azure/src/lib/ConsumptionTypes.ts), they're "`50000 GB Seconds`" & "`1000 GB Hours`".
```
export enum MEMORY_USAGE_UNITS {
  GB_SECONDS_50000 = '50000 GB Seconds',
  GB_HOURS_1000 = '1000 GB Hours',
}
```

Do you have any plan to support "`1 GB second`" and "`1GB hour`" unit?

Here's the Google groups conversation link: [Unsupported usage unit of Azure](https://groups.google.com/g/cloud-carbon-footprint/c/cb8xhFt27v8)

Thanks and Regards,
Hank

**Comment by 4upz** [comment]
Hi @HankLeo thanks for submitting this issue! We'd definitely want to add support for those usage units, so we appreciate you sharing your warnings along with the usage examples as that helps us constantly improve support for the various usage units that can be returned from different billing data.

Adding support will involve adding the `1 GB Hour` and `1 GB Second` units to the `MEMORY_USAGE_UNITS` object in the ConsumptionTypes files. Afterwards, a case for these units will need to be added to the [ConsumptionManagement](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L498C7-L499) file's pricing unit function. The [`getUsageAmountInGigabyteHours`](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L752) function will also need to be updated to handle converting `1 GB Second` to GB Hours. 

We'll add this to our backlog to implement. However, if those steps make sense and if this is something you'd be interested in taking a stab at contributing, just let us know and we'd be happy to support!

**Comment by HankLeo** [comment]
> Adding support will involve adding the `1 GB Hour` and `1 GB Second` units to the `MEMORY_USAGE_UNITS` object in the ConsumptionTypes files. Afterwards, a case for these units will need to be added to the [ConsumptionManagement](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L498C7-L499) file's pricing unit function. The [`getUsageAmountInGigabyteHours`](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L752) function will also need to be updated to handle converting `1 GB Second` to GB Hours.

Hi @4upz ,
Thanks for your reply for this question, and the solution looks great to me! I review the codes you mentioned, but I'm not very clear about the logic of the [`getUsageAmountInGigabyteHours`](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L752) function. What's the reason that the GB Seconds usageUnit just divides 3600 (seconds/hour) directly but GB Hours usageUnit should run the function [`getGigabyteHoursFromInstanceTypeAndProcessors`](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L873)? And by following the current logic, seems I need add below 2 conditions in [`getUsageAmountInGigabyteHours`](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L752) function, am I right?
```
private getUsageAmountInGigabyteHours(
  consumptionDetailRow: ConsumptionDetailRow,
): number {
  if (consumptionDetailRow.usageUnit === MEMORY_USAGE_UNITS.GB_SECONDS_50000)
    return consumptionDetailRow.usageAmount / 3600
  if (consumptionDetailRow.usageUnit === MEMORY_USAGE_UNITS.GB_SECOND_1)
    return consumptionDetailRow.usageAmount / 3600 / 50000
  if (consumptionDetailRow.usageUnit === MEMORY_USAGE_UNITS.GB_HOUR_1)
    consumptionDetailRow.usageAmount /= 1000
  return this.getGigabyteHoursFromInstanceTypeAndProcessors(
    consumptionDetailRow.usageType,
    consumptionDetailRow.usageAmount,
    consumptionDetailRow.seriesName,
  )
}
```

**Comment by 4upz** [comment]
Hi @HankLeo,

Sure, I'd be happy to explain! The purpose of the `getUsageAmountInGigabyteHours` function is to convert the methods of usage units into a value that represents usage in terms of gigabytes per hour. Since there are 3600 seconds in an hour, it divides the usageAmount by that value if the usage unit is reported in seconds. 

It does not do this for `MEMORY_USAGE_UNITS.GB_HOURS_1000` since this unit is already in GBs and Hours. It can simply return the usage amount as-is, however, it makes the function call to ensure that the `1000 GB Hours` unit isn't associated with a [memory cache](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/056c0c8b0ab0910ed7a613fc6814fd5ae6163ed3/packages/azure/src/lib/ConsumptionManagement.ts#L878-L883).

You're right in that you'll need to add the extra conditions, so your example looks good! However, the only important thing is converting the units. No need to convert the values to fit a certain GB/Hr rate as the usageAmount will already reflect that. Therefore, I don't think you need to divide by an extra 50000 for `MEMORY_USAGE_UNITS.GB_SECOND_1`. For `MEMORY_USAGE_UNITS.GB_HOUR_1` you can just return the usage amount. No need to divide by 1000 or reassign the usageAmount property.

So in short, it would look like this:
``` typescript
private getUsageAmountInGigabyteHours(
  consumptionDetailRow: ConsumptionDetailRow,
): number {
  if (consumptionDetailRow.usageUnit === MEMORY_USAGE_UNITS.GB_SECONDS_50000 || consumptionDetailRow.usageUnit === MEMORY_USAGE_UNITS.GB_SECOND_1)
    return consumptionDetailRow.usageAmount / 3600
  if (consumptionDetailRow.usageUnit === MEMORY_USAGE_UNITS.GB_HOUR_1)
    return consumptionDetailRow.usageAmount
  return this.getGigabyteHoursFromInstanceTypeAndProcessors(
    consumptionDetailRow.usageType,
    consumptionDetailRow.usageAmount,
    consumptionDetailRow.seriesName,
  )
}
```

I hope that makes sense! Let me know if you have any other questions or need additional support.

**Comment by HankLeo** [comment]
Hi @4upz ,
Thanks for your explanation. And I also submit the PR as linked. It's my first time to submit the PR in this project, not sure if I missed following any guidance. Please let me know if I need do further change, I'm glad to contribute this project😉.

**Comment by 4upz** [comment]
@HankLeo Awesome, thank you for opening the PR and congrats on your first one! I'll give it a review and will leave any additional feedback there
